phd-tests/tests/src/stats.rs

```diff
@@ -0,0 +1,495 @@
+    // below.
+    let idle_start_metrics =
+        fake_nexus.app_private().wait_for_propolis_stats().await;
+    let idle_start = std::time::SystemTime::now();
+    source.run_shell_command("sleep 20").await?;
+
+    let now_metrics = fake_nexus.app_private().wait_for_propolis_stats().await;
+
+    // The guest VM would continues to exist with its idle vCPU being accounted
+    // by the kstats Oximeter samples. This means `wait_for_propolis_stats`
+    // could introduce as much as a full Oximeter sample interval of additional
+    // idle vCPU, and is we why wait to measure idle time until *after* getting
+    // new Oximeter metrics.
+    let idle_time = idle_start.elapsed().expect("time goes forwards");
+    trace!("measured idle time {:?}", idle_time);
+
+    let idle_delta = (now_metrics.vcpu_state_total(&VcpuState::Idle)
+        - idle_start_metrics.vcpu_state_total(&VcpuState::Idle))
+        as u128;
+
+    // We've idled for at least 20 seconds. The guest may not be fully idle (its
+    // OS is still running on its sole CPU, for example), so we test that the
+    // guest was just mostly idle for the time period.
+    let min_guest_idle_delta = (idle_time.as_nanos() as f64 * 0.9) as u128;
+    assert!(
+        idle_delta < idle_time.as_nanos(),
+        "{} < {}",
+        idle_delta as f64 / NANOS_PER_SEC,
+        idle_time.as_nanos() as f64 / NANOS_PER_SEC
+    );
+    assert!(
+        idle_delta > min_guest_idle_delta,
+        "{} > {}",
+        idle_delta as f64 / NANOS_PER_SEC,
+        min_guest_idle_delta as f64 / NANOS_PER_SEC
+    );
+
+    // The delta in vCPU `run` time should be negligible. We've run one shell
+    // command which in turn just idled.
+    let run_delta = (now_metrics.vcpu_state_total(&VcpuState::Run)
+        - idle_start_metrics.vcpu_state_total(&VcpuState::Run))
+        as u128;
+    assert!(run_delta < Duration::from_millis(100).as_nanos());
+
+    let full_run_delta = (now_metrics.vcpu_state_total(&VcpuState::Run)
+        - start_metrics.vcpu_state_total(&VcpuState::Run))
+        as u128;
+
+    let full_idle_delta = (now_metrics.vcpu_state_total(&VcpuState::Idle)
+        - start_metrics.vcpu_state_total(&VcpuState::Idle))
+        as u128;
+
+    let full_waiting_delta = (now_metrics.vcpu_state_total(&VcpuState::Waiting)
+        - start_metrics.vcpu_state_total(&VcpuState::Waiting))
+        as u128;
+
+    let full_emul_delta = (now_metrics.vcpu_state_total(&VcpuState::Emulation)
+        - start_metrics.vcpu_state_total(&VcpuState::Emulation))
+        as u128;
+
+    // Pick 100ms as a comically high upper bound for how much time might have
+    // been spent emulating instructions on the guest's behalf. Anecdotally the
+    // this is on the order of 8ms between the two samples. This should be very
+    // low; the workload is almost entirely guest user mode execution.
+    assert!(full_emul_delta < Duration::from_millis(100).as_nanos());
```

Comment:
this is basically the whole motivator for the test. we'd seen a chart where `emul` seemed to track closely but not perfectly with `run`, and now we can be a bit more confident that the system is producing coherent numbers at least, and what we saw there was an issue somewhere further up.

Reply:
Forgive the naive question, by why would one expect `emul` to perfectly track `run`?

Reply:
ah, i wouldn't expect `emul` to perfectly track `run` either. if they were exactly the same, i would have just first looked for something to be mislabeled (anywhere between illumos and the console)

close-but-not-perfectly tells me that it's not just mislabeled data, which could be illumos calculating stats in a strange way, or something weird about the sampling, or something later at query time. so, test helps rule out the first two possibilities..

my actual expectation is that `emul` might count up a little more often when `run` is counting up, but as a side effect of having exit for instruction emulation. they ought to be orders of magnitude different relatively independent.

Reply:
Thanks, that's helpful. I misunderstood your first comment as meaning that you expected run / emul to be very highly correlated. It sounds like you expect a weak positive correlation, which makes sense to me. Is the issue is that you see the values being much closer in absolute terms than you expect? How close are we talking about here?

I'm just curious about all this, I have very little intuition about what one _should_ see.